rudimentary sonos support

rhooper opened the pull request:
Basic sonos media_player support via the SoCo sonos module.

rhooper committed:
rudimentary sonos support

rhooper committed:
fix initialization

rhooper committed:
tidy up formatting to make travis happy.

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -0,0 +1,194 @@
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-public-methods
+class SonosDevice(MediaPlayerDevice):
+    """ Represents a SqueezeBox device. """
```
Documentation out of date

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -0,0 +1,194 @@
+    def turn_off(self):
+        """ turn_off media player. """
+        self._player.pause()
+        self.update_ha_state()
```
This one is not needed, media player component will take care of this.

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -0,0 +1,194 @@
+        self._player.mute = mute
+        self.update_ha_state()
+
+    def media_play_pause(self):
```
This method is not needed, it is [implemented](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L469-L474) like this in the base class.

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -0,0 +1,194 @@
+        self._player.play()
+        self.update_ha_state()
+
+    def play_youtube(self, media_id):
```
This one is not needed either

rhooper committed:
Enable polling and fix metadata updating.
Remove unnecessary methods.
Include SoCo in requirements_all.txt for CI.
Lock down SoCo version to 0.11.1
Add sonos.py to exclusions in .coveragerc

rhooper committed:
put back play_youtube override

rhooper committed:
remove and disable pylint: disable=abstract-method for play_youtube()

rhooper committed:
remove unnecessary self.update_ha_state calls

rhooper committed:
use own track_utc_time_change to poll every 5 seconds

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -0,0 +1,180 @@
+        self._player = player
+        self.update()
+
+        track_utc_time_change(
```
The fact that this works is a matter of luck ;-)

Tracking a time changed event will call `callback(now)` with `now` being a datetime object. `update_ha_state` takes 1 argument being `force_update`. If true-ish, it will call `update()`.

The API probably won't change but I prefer if you would do this:
```python
def update_sonos(now):
    self.update_ha_state(True)

track_utc_time_change(…)
```

I prefer a method like this over a lambda so that when the job queue gets big, it has a logical name.

rhooper committed:
follow proper calling convention for track_utc_time_change callback
